connectors/src/lib/data_sources.ts

```diff
@@ -111,15 +130,31 @@ async function _upsertToDatasource({
     );
   } catch (e) {
     const elapsed = new Date().getTime() - now.getTime();
-    if (axios.isAxiosError(e) && e.config?.data) {
-      e.config.data = "[REDACTED]";
-    }
+
     statsDClient.increment("data_source_upserts_error.count", 1, statsDTags);
     statsDClient.distribution(
       "data_source_upserts_error.duration.distribution",
       elapsed,
       statsDTags
     );
+
+    if (axios.isAxiosError(e) && e.config?.data) {
+      e.config.data = "[REDACTED]";
+
+      switch (e.status) {
+        case 401: {
+          const dataSourceQuotaError = DataSourceQuotaError.fromAxiosError(e);
+
+          localLogger.error(
+            { error: dataSourceQuotaError },
+            "Error uploading document to Dust."
+          );
+
+          throw dataSourceQuotaError;
```

Comment:
Technically this is not aligned with our error handling principles.

we should return a Result here if we plan to act on it as per current invariants.

Even better we should move to using DustAPI here as well 😬